plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,241 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.allOf;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class FileSystemExchangeSink
+        implements ExchangeSink
+{
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FileSystemExchangeSink.class).instanceSize();
+    private static final int INTEGER_INSTANCE_SIZE = ClassLayout.parseClass(Integer.class).instanceSize();
+
+    private final FileSystemExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final int numBuffers;
+    private final Optional<SecretKey> secretKey;
+    private final int writeBufferSize;
+
+    private final Map<Integer, ExchangeStorageWriter> writers = new ConcurrentHashMap<>();
+    private final Map<Integer, SliceOutput> pendingBuffers = new ConcurrentHashMap<>();
+    private final LinkedBlockingQueue<SliceOutput> freeBuffers = new LinkedBlockingQueue<>();
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+    private volatile boolean finished;
+    @GuardedBy("this")
+    private CompletableFuture<Void> blockedFuture = new CompletableFuture<>();
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.numBuffers = outputPartitionCount * 2; // double buffering to overlap computation and I/O
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        this.writeBufferSize = exchangeStorage.getWriteBufferSizeInBytes();
+
+        for (int i = 0; i < numBuffers; ++i) {
+            freeBuffers.add(Slices.allocate(writeBufferSize).getOutput());
+        }
+    }
+
+    // The future returned by {@link #isBlocked()} should be considered as a best-effort hint. It's
+    // possible to return NOT_BLOCKED when it's actually blocked in certain cases. The decision is
+    // consciously made to avoid global locking.
+    @Override
+    public CompletableFuture<?> isBlocked()
+    {
+        if (freeBuffers.isEmpty() && pendingBuffers.size() < outputPartitionCount) {
```

Comment:
I'm fine both ways. `pendingBuffers.size() == outputPartitionCount` means that for every partition we still have some room to buffer incoming data. It can become blocked or not blocked, as we don't know the remaining buffer size and the incoming slice size